File: bench/kernels/e2_checksum/ref.c
/* e2-checksum (family E), naive C. Multiply-add-mask rolling hash; signed
 * overflow is UB so clang need not prove anything.
 * Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static int64_t checksum(int64_t n, int64_t seed) {
    int64_t h = seed;
    for (int64_t i = 0; i < n; i++) h = (h * 31 + (i & 255)) & 1048575;
    return h;
}

int main(int argc, char **argv) {
    int64_t ops = argc > 1 ? (int64_t)strtoull(argv[1], 0, 10) : 4000;
    const int64_t inner = 100000;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    int64_t sink = 1;
    for (int64_t k = 0; k < ops; k++) sink = checksum(inner, sink);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%lld,\"sink\":%lld}\n",
           (unsigned long long)ns, (long long)(ops * inner), (long long)sink);
    return 0;
}

File: bench/kernels/e2_checksum/ref.rs
//! e2-checksum (family E), Rust reference.
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;

fn checksum(n: i64, seed: i64) -> i64 {
    let mut h: i64 = seed;
    let mut i: i64 = 0;
    while i < n {
        h = (h * 31 + (i & 255)) & 1048575;
        i += 1;
    }
    h
}

fn main() {
    let ops: i64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(4000);
    let inner: i64 = 100_000;
    let t0 = std::time::Instant::now();
    let mut sink: i64 = 1;
    for _ in 0..ops {
        sink = checksum(black_box(inner), sink);
    }
    let ns = t0.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops.wrapping_mul(inner)
    );
}
